wicket/src/cli/rack_update.rs

```diff
@@ -380,6 +404,155 @@ async fn do_clear_update_state(
+
+    #[clap(flatten)]
+    component_ids: ComponentIdSelector,
+}
+
+impl ReplayArgs {
+    fn exec(
+        self,
+        log: Logger,
+        global_opts: GlobalOpts,
+        output: CommandOutput<'_>,
+    ) -> Result<()> {
+        let update_ids = self.component_ids.to_component_ids()?;
+        let mut display = GroupDisplay::new_with_display(
+            &log,
+            update_ids.iter().copied(),
+            output.stderr,
+        );
+        if global_opts.use_color() {
+            display.set_styles(LineDisplayStyles::colorized());
+        }
+
+        let file = BufReader::new(
+            std::fs::File::open(&self.file)
+                .with_context(|| format!("error opening {}", self.file))?,
+        );
+        let response: GetArtifactsAndEventReportsResponse =
+            serde_json::from_reader(file)?;
+        let event_reports =
+            parse_event_report_map(&log, response.event_reports);
+
+        self.strategy.execute(display, event_reports)?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Default, Eq, PartialEq, Hash, Debug, ValueEnum)]
+enum ReplayStrategy {
+    /// Feed all events into the buffer immediately.
+    #[default]
+    Oneshot,
+
+    /// Feed events into the buffer one at a time.
+    Incremental,
+
+    /// Feed events into the buffer as 0, 0..1, 0..2, 0..3 etc.
+    Idempotent,
+}
+
+impl ReplayStrategy {
+    fn execute(
+        self,
+        mut display: GroupDisplay<
+            ComponentId,
+            &mut dyn Write,
+            WicketdEngineSpec,
+        >,
+        event_reports: BTreeMap<ComponentId, EventReport>,
+    ) -> Result<()> {
+        match self {
+            ReplayStrategy::Oneshot => {
+                // TODO: parallelize this computation?
+                for (id, event_report) in event_reports {
+                    // If display.add_event_report errors out, it's for a report for a
+                    // component we weren't interested in. Ignore it.
+                    _ = display.add_event_report(&id, event_report);
+                }
+
+                display.write_events()?;
+            }
+            ReplayStrategy::Incremental => {
+                for (id, event_report) in &event_reports {
+                    let mut buffer = EventBuffer::default();
+                    let mut last_seen = None;
+                    for event in &event_report.step_events {
+                        buffer.add_step_event(event.clone());
+                        let report =
+                            buffer.generate_report_since(&mut last_seen);
+
+                        // If display.add_event_report errors out, it's for a report for a
+                        // component we weren't interested in. Ignore it.
+                        _ = display.add_event_report(&id, report);
+
+                        display.write_events()?;
+                    }
+                }
+            }
+            ReplayStrategy::Idempotent => {
```

Comment:
This is the best name I could come up with for this strategy -- would love ideas for better names.

Reply:
`Idempotent` seems fine to me.  The only alternative I could think of is `Full` or `Complete`.

Reply:
I'd considered "full", but I think it isn't clearly differentiated from "oneshot" -- "idempotent" is slightly weird at first but makes sense when you see what it's doing.